trajopt_ifopt/src/cartesian_line_constraint.cpp

```diff
@@ -0,0 +1,222 @@
+ *
+ * @par License
+ * Software License Agreement (Apache License)
+ * @par
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * @par
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <trajopt_ifopt/constraints/cartesian_line_constraint.h>
+#include <trajopt_ifopt/utils/numeric_differentiation.h>
+
+TRAJOPT_IGNORE_WARNINGS_PUSH
+#include <tesseract_kinematics/core/utils.h>
+#include <console_bridge/console.h>
+TRAJOPT_IGNORE_WARNINGS_POP
+
+namespace trajopt
+{
+CartLineConstraint::CartLineConstraint(const Eigen::Isometry3d& origin_pose,
+                                       const Eigen::Isometry3d& target_pose,
+                                       CartLineKinematicInfo::ConstPtr kinematic_info,
+                                       JointPosition::ConstPtr position_var,
+                                       const std::string& name)
+  : ifopt::ConstraintSet(6, name)
+  , position_var_(std::move(position_var))
+  , point_a_(origin_pose)
+  , point_b_(target_pose)
+  , kinematic_info_(std::move(kinematic_info))
+{
+  // Set the n_dof and n_vars for convenience
+  n_dof_ = kinematic_info_->manip->numJoints();
+  assert(n_dof_ > 0);
+
+  bounds_ = std::vector<ifopt::Bounds>(6, ifopt::BoundZero);
+
+  // calculate the equation of the constraint line
+  line_ = point_b_.translation() - point_a_.translation();
+}
+
+Eigen::VectorXd CartLineConstraint::CalcValues(const Eigen::Ref<const Eigen::VectorXd>& joint_vals) const
+{
+  Eigen::Isometry3d new_pose = kinematic_info_->manip->calcFwdKin(joint_vals);
+
+  new_pose = kinematic_info_->world_to_base * new_pose * kinematic_info_->kin_link->transform * kinematic_info_->tcp;
+
+  // For Jacobian Calc, we need the inverse of the nearest point, D, to new Pose, C, on the constraint line AB
+  // This will not honor tcp orientation.
+
+  // d1 = distance 1; distance from new pose to first point on line
+  Eigen::Vector3d d1 = (point_a_.translation() - new_pose.translation()).array().abs();
+
+  // Point D, the nearest point on line AB to point C, can be found with:
+  // (AC - (AC * AB)) * AB
+  Eigen::Vector3d line_point;
+  Eigen::Vector3d line_norm = line_ / line_.squaredNorm();
+  double mag = d1.dot(line_norm);
+
+  // If point C is not between the line endpoints, set nearest point to endpoint
+  if (mag > 1.0)
+  {
+    line_point = point_b_.translation();
+  }
+  else if (mag < 0)
+  {
+    line_point = point_a_.translation();
+  }
+  else
+  {
+    line_point = point_a_.translation() + mag * line_norm;
```

Comment:
Should line_point_ be stored here?

Reply:
It should. However, this class extends the ifopt::ConstraintSet. The CalcValues function that is inherited from this object is a const method. To preserve this, I could not save off the line_point here as I intended.